Fix: Support class-level @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY) on custom Collection subclasses

jher235 opened the pull request:
### Issue
Closes #5522 
### 1. What does this PR do?

This PR extends `CollectionDeserializer` to respect `@JsonFormat` annotations declared directly on custom `Collection` subclasses.

Currently, `CollectionDeserializer` ignores class-level annotations (e.g., `@JsonFormat class MyList ...`) and only considers property-level configurations or global `ConfigOverride`. This PR adds a **fallback mechanism** to explicitly check the container type's annotations when no property-level configuration is found.

### 2. Why is this change necessary?

**A. Addressing Feature Inconsistency**
While `@JsonFormat` is allowed on classes, `CollectionDeserializer` currently ignores it for enabling single-value uwrapping. This aligns the behavior with other types (like POJOs) where class-level features are respected.

**B. Enabling Fine-Grained Control**
Users currently rely on `mapper.configOverride(Collection.class)` which affects **ALL** collections globally.
This change allows users to enable `ACCEPT_SINGLE_VALUE_AS_ARRAY` for specific collection types only:

```java
// Now supported:
@JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
public class FlexibleList<E> extends ArrayList<E> { }

class MyDTO {
    private FlexibleList<String> items;
}
```

### 3. Implementation Details

I implemented this as a **fallback** to ensure full backward compatibility.

**Resolution Order:**

1. **Property Annotation** (Highest priority, unchanged)
2. **ConfigOverride for `Collection.class`** (Existing global config, unchanged)
3. **Concrete Container Class Annotation** (New fallback logic for `ACCEPT_SINGLE_VALUE_AS_ARRAY`)
4. **Global DeserializationFeature**

Note on Safety:

I purposely maintained the existing findFormatFeature(..., Collection.class) call as the primary check. Replacing Collection.class with the concrete type would break existing ConfigOverride settings for generic collections. The new check is strictly additive.

### 4. Test Coverage

Added `CollectionDeserializer5522Test` covering:

- `@JsonFormat` on custom collection subclass is respected.
- Property-level annotation still takes precedence.
- Standard `Collection` behavior remains unaffected.

jher235 committed:
Verify do not ignore @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY) in custom Collections subclasses

Signed-off-by: jher235 <tim668666@gmail.com>

jher235 committed:
Add fallback logic to support @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY) on custom Collections subclasses.

Signed-off-by: jher235 <tim668666@gmail.com>

pjfanning commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -172,6 +172,13 @@ public CollectionDeserializer createContextual(DeserializationContext ctxt,
         Boolean unwrapSingle = findFormatFeature(ctxt, property, Collection.class,
                 JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
+        if (unwrapSingle == null) {
+            tools.jackson.databind.introspect.AnnotatedClass ac = ctxt.introspectClassAnnotations(_containerType);
```
Can you import this class instead of fully referencing it here?

Can you create a static method that StringCollectionDeserializer.java‎ can use too instead of duplicating the code in StringCollectionDeserializer? Or if there is a shared superclass, you could put the method there.

jher235 replied:
Sure, I will fix the import. Regarding the duplication, I will extract the added logic into a static method in CollectionDeserializer and have StringCollectionDeserializer call it.

pjfanning commented on src/test/java/tools/jackson/databind/deser/jdk/CollectionDeserializer5522Test.java:
```diff
@@ -0,0 +1,116 @@
+            .containsExactly("test");
+    }
+
+}
```
new line at end of file please

JooHyukKim commented:
@jher235 Can we also have another test case with Pojo-wrapping-CustomNumberList? Currently CustomNumberList is at root-level. 

Wonder how behavior might change...

jher235 commented:
Hello @JooHyukKim 

Current test case `testCustomNumberCollectionDeserialize` actually uses `CustomClassForNumber`, which is a POJO wrapper containing `CustomNumberList` as a field (`value`).
``` java 
static class CustomClassForNumber {
    private CustomNumberList value;
    ...
}
```
Is this the "Pojo-wrapping" test you were looking for?
Or would you like me to add a "Root-level" test (deserializing CustomNumberList directly) as well?

jher235 committed:
Extract fallback logic to static method on CollectionDeserializer

Signed-off-by: jher235 <tim668666@gmail.com>

jher235 committed:
Add new line at end of file

Signed-off-by: jher235 <tim668666@gmail.com>

pjfanning commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -602,4 +598,18 @@ public void handleResolvedForwardReference(DeserializationContext ctxt, Object i
         }
     }
+
+    public static Boolean findFormatFeatureOrClassFallback(DeserializationContext ctxt,
```
remove 'public', package private is fine

jher235 committed:
Remove access modifier at findFormatFeatureOrClassFallback method

Signed-off-by: jher235 <tim668666@gmail.com>

cowtowncoder commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -172,6 +173,8 @@ public CollectionDeserializer createContextual(DeserializationContext ctxt,
         //   comes down to "List vs Collection" I suppose... for now, pass Collection
         Boolean unwrapSingle = findFormatFeature(ctxt, property, Collection.class,
                 JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
+        unwrapSingle = findFormatFeatureOrClassFallback(ctxt, _containerType, unwrapSingle,
```
Ok, logic makes sense, but I think precedence is not quite right since the new value has lower precedence than everything else. It should be in the middle; from highest to lowest:

1. Property-override
2. Collection class annotation (new thing added here)
3. ConfigOverride for `Collection`
4. Global `JsonFormat` via ConfigOverride

But maybe what is here is good enough... I'll have to think about it.

cowtowncoder commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -595,4 +598,18 @@ public void handleResolvedForwardReference(DeserializationContext ctxt, Object i
         }
     }
+
+    static Boolean findFormatFeatureOrClassFallback(DeserializationContext ctxt,
```
Change to non-static and don't pass `_containerType`, refer from here.

cowtowncoder commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -595,4 +598,18 @@ public void handleResolvedForwardReference(DeserializationContext ctxt, Object i
+        if (unwrapSingle != null) {
+            return unwrapSingle;
+        }
+        AnnotatedClass ac = ctxt.introspectClassAnnotations(type);
```
Ideally we would not need to re-introspect (adds to startup overhead)... since `BasicDeserializerFactory.createCollectionDeserializer(...)` has `BeanDescription.Supplier beanDescRef` and could pass it to `CollectionDeserializer`.

That'd require a bit of refactoring that might be worth it: instead of direct call to constructor, should call factory method... I'll go ahead and do that part of refactoring now.

cowtowncoder replied:
Ok, #5525 added 

```
    public static CollectionDeserializer create(JavaType collectionType,
            BeanDescription.Supplier beanDescRef,
            ValueDeserializer<Object> valueDeser,
            TypeDeserializer valueTypeDeser, ValueInstantiator valueInstantiator)

```

where

    beanDescRef.getClassAnnotations();

can be used to get introspected class annotations of target type. This could be passed to constructor (new overload with added parameter), retained in a `transient` field, to be available for `createContextual()`. May sound like lots of work but annotation introspection is becoming bigger and bigger problem with start up overhead with Jackson so I'd like to limit additional calls to re-introspecting again and again.

jher235 committed:
Merge remote-tracking branch 'upstream/3.x' into fix/5522-support-jsonformat-on-custom-collection

jher235 committed:
Cache class annotations to avoid re-introspection in Collection deserializers

Signed-off-by: jher235 <tim668666@gmail.com>

jher235 commented:
Hello @cowtowncoder  
Thanks for the detailed review!

I implemented the logic as non-static instance methods in both classes as suggested. I decided to keep them as separate helper methods (findFormatFeatureOrClassFallback) instead of inlining them into createContextual, as it keeps the main logic much cleaner.
Also, I set the helper method to protected in CollectionDeserializer (for extensibility) but private in StringCollectionDeserializer since it is a final class.

cowtowncoder commented:
+1 for separate lookup methods instead of inlining. Static vs non-static matter of preference, not a big deal for me.

But let's see if we can avoid re-introspecting class annotations, as per my suggestion.

jher235 commented:
Actually, I did it in my latest commit `9935f3d`. Could you please take a look?

cowtowncoder commented on src/main/java/tools/jackson/databind/deser/jdk/CollectionDeserializer.java:
```diff
@@ -217,6 +228,21 @@ public CollectionDeserializer createContextual(DeserializationContext ctxt,
+            return unwrapSingle;
+        }
+        if(_classAnnotations != null){
+            JsonFormat formatAnnotation = _classAnnotations.get(JsonFormat.class);
```
Right idea, but implementation needs bit more work: the only code that should ever access concrete annotations is `[Jackson]AnnotationIntrospector`. So we need to hold on to`AnnotatedClass`, pass that to `AnnotationIntrospector.findFormat()`.

(the reason `AnnotationIntrospector` needs to be involved is mainly for 2 things:
1. mix-in annotations
2. non-Jackson (JAXB) annotation suppor
but also gives a powerful extension point for overrides).

cowtowncoder committed:
Merge branch '3.x' into fix/5522-support-jsonformat-on-custom-collection

cowtowncoder commented:
> Actually, I did it in my latest commit `9935f3d`. Could you please take a look?

Nice! Much better, just needs work to get `AnnotationIntrospector` involved (as per my inline note), instead of direct annotation lookup.

jher235 committed:
Refactored Collection deserializers to use AnnotatedClass instead of Annotations
This ensures that Mix-ins, JAXB annotations, and configuration overrides are correctly respected while avoiding re-introspection overhead.

Signed-off-by: jher235 <tim668666@gmail.com>

jher235 commented:
@cowtowncoder 
I refactored Collection deserializers to use `AnnotatedClass` instead of `Annotations` following your feedback.

cowtowncoder commented:
Thank you @jher235 -- I hope to review this within couple of days (on christmas vacation).

cowtowncoder committed:
Merge branch '3.x' into fix/5522-support-jsonformat-on-custom-collection

cowtowncoder committed:
Merge branch '3.x' into fix/5522-support-jsonformat-on-custom-collection

cowtowncoder committed:
Add back Constructor as deprecated for better backwards-compatibility

cowtowncoder committed:
Merge branch '3.x' into fix/5522-support-jsonformat-on-custom-collection

cowtowncoder committed:
Refactoring; add release notes

cowtowncoder committed:
Minor javadoc/comment clean up
